python_modules/dagster/dagster/core/definitions/job_definition.py

```diff
@@ -65,6 +68,9 @@ def __init__(
         hook_defs: Optional[AbstractSet[HookDefinition]] = None,
         op_retry_policy: Optional[RetryPolicy] = None,
         version_strategy: Optional[VersionStrategy] = None,
+        asset_key_by_output_def: Optional[Mapping[OutputDefinition, AssetKey]] = None,
```

Comment:
I think making the `OutputDefinition` the key here might cause trouble because nothing stops two assets from sharing the same output def.

E.g.
```
@op
def my_op():
    ...

assets_def1 = AssetsDefinition(
    op=my_op,
    asset_keys_by_output_name={"result": AssetKey("abc")},
    asset_keys_by_input_name={},
)

assets_def2 = AssetsDefinition(
    op=my_op,
    asset_keys_by_output_name={"result": AssetKey("xyz")},
    asset_keys_by_input_name={},
)

my_job = AssetGroup([assets_def1, assets_def2]).build_job("fsdjk")
```

Reply:
yeah good point -- after playing around with things a bit, our current implementation also breaks with this sort of setup (for different reasons), so I'll fix that as well.